File: detector/src/kitsune_detector/coherence/engine.py
# detector/coherence/engine — generic evaluator over data-driven rules.
# Resolves each rule's reads against a session and emits contradictions.

"""The coherence engine — a generic evaluator over data-driven rules.

The engine knows *how* to evaluate (resolve a rule's reads against a session, run its predicate,
emit a contradiction). It knows nothing about *which* incoherences matter — that is the registry's
job. This separation is what keeps Kitsune maintainable as signals decay.
"""

from __future__ import annotations

from ..models import Contradiction, Layer, Session
from .predicates import PREDICATES
from .rules import CoherenceRule, RuleSet


class CoherenceEngine:
    def __init__(self, ruleset: RuleSet) -> None:
        self._ruleset = ruleset

    @property
    def ruleset_version(self) -> str:
        return self._ruleset.ruleset_version

    @staticmethod
    def _resolve(session: Session, ref: str) -> object:
        """Resolve a ``layer.kind`` read to its signal value, or MISSING."""
        layer_name, _, kind = ref.partition(".")
        return session.value(Layer(layer_name), kind)

    def _fire(self, session: Session, rule: CoherenceRule) -> bool:
        predicate = PREDICATES[rule.predicate]
        values = [self._resolve(session, ref) for ref in rule.reads]
        return predicate(values, rule.threshold)

    def evaluate(self, session: Session) -> list[Contradiction]:
        """Return every contradiction that fires for this session."""
        contradictions: list[Contradiction] = []
        for rule in self._ruleset.evaluable_rules:
            if self._fire(session, rule):
                contradictions.append(
                    Contradiction(
                        rule_id=rule.id,
                        layers=rule.layers,
                        detail=rule.title,
                        weight=rule.weight,
                        evidence=list(rule.reads),
                    )
                )
        return contradictions

File: collector/src/livepage/engine.ts
// collector/livepage/engine — evaluate the rule registry against collected signals, in the browser.
// Mirrors detector/coherence/engine.py: resolve each rule's reads, fire its predicate, score a verdict.

import type { Layer, SignalValue } from "../types.js";
import { MISSING, PREDICATES, type Resolved } from "./predicates.js";
import type { RuleJSON } from "./registry.js";
import {
  finalScore,
  incoherenceScore,
  type Label,
  labelFor,
  layerScores,
  type LayerScores,
} from "./scoring.js";

/** A signal map keyed by "layer.kind" — the same reference form rules read. */
export type SignalMap = Map<string, SignalValue>;

export interface Contradiction {
  id: string;
  title: string;
  layers: Layer[];
  weight: number;
  category: string;
  /** The "layer.kind" references whose values fired the rule. */
  evidence: string[];
}

function resolve(signals: SignalMap, ref: string): Resolved {
  return signals.has(ref) ? (signals.get(ref) as SignalValue) : MISSING;
}

/** Fire every rule whose predicate holds over its resolved reads — returns the contradictions. */
export function evaluate(rules: RuleJSON[], signals: SignalMap): Contradiction[] {
  const out: Contradiction[] = [];
  for (const rule of rules) {
    const values = rule.reads.map((ref) => resolve(signals, ref));
    if (PREDICATES[rule.predicate](values, rule.threshold)) {
      out.push({
        id: rule.id,
        title: rule.title,
        layers: rule.layers,
        weight: rule.weight,
        category: rule.category,
        evidence: rule.reads,
      });
    }
  }
  return out;
}

export interface Verdict {
  score: number;
  label: Label;
  incoherence: number;
  layers: LayerScores;
  contradictions: Contradiction[];
}

export function verdictFor(contradictions: Contradiction[]): Verdict {
  const score = finalScore(contradictions);
  return {
    score,
    label: labelFor(score, contradictions), // conviction-gated: corroborating-only tells cap at suspicious
    incoherence: incoherenceScore(contradictions),
    layers: layerScores(contradictions),
    contradictions,
  };
}
